gax-httpjson/src/main/java/com/google/api/gax/httpjson/ProtoRestSerializer.java

```diff
@@ -127,16 +169,34 @@ public void putPathParam(Map<String, String> fields, String fieldName, Object fi
    * @param fieldValue a field value
    */
   public void putQueryParam(Map<String, List<String>> fields, String fieldName, Object fieldValue) {
-    ImmutableList.Builder<String> paramValueList = ImmutableList.builder();
+    ArrayList<String> paramValueList = new ArrayList();
     if (fieldValue instanceof List<?>) {
+      boolean hasProcessedMessage = false;
       for (Object fieldValueItem : (List<?>) fieldValue) {
-        paramValueList.add(String.valueOf(fieldValueItem));
+        if (isNonSerializableMessageValue(fieldValueItem)) {
+          putDecomposedMessageQueryParam(fields, fieldName, fieldValueItem);
+          hasProcessedMessage = true;
+        } else {
+          paramValueList.add(toQueryParamValue(fieldValueItem));
+        }
+      }
+      if (hasProcessedMessage) {
```

Comment:
Rather than introducing this variable and a secondary exit point for the function, consider instead just performing a `!paramValueList.isEmpty()` check at the end to decide whether to add to the final `fields` map or not.